sql/catalyst/src/main/java/org/apache/spark/sql/connector/expressions/Expression.java

```diff
@@ -44,7 +44,16 @@ public interface Expression {
    * List of fields or columns that are referenced by this expression.
    */
   default NamedReference[] references() {
-    return Arrays.stream(children()).map(e -> e.references())
-      .flatMap(Arrays::stream).distinct().toArray(NamedReference[]::new);
+    List<NamedReference> list = new ArrayList<>();
+    Set<NamedReference> uniqueValues = new HashSet<>();
+    for (Expression e : children()) {
+      NamedReference[] references = e.references();
+      for (NamedReference reference : references) {
+        if (uniqueValues.add(reference)) {
+          list.add(reference);
+        }
+      }
+    }
+    return list.toArray(new NamedReference[0]);
```

Comment:
Do you need to build the List too? why not just .toArray on the Set, because ordering is important? LinkedHashSet could help there

Reply:
let met check this

Reply:
[1ecf017](https://github.com/apache/spark/pull/37843/commits/1ecf0172b2065c250ddbc44f5d0e6d00b7f73895) change to use LinkedHashSet, let me check the performance and waiting CI

Reply:
@srowen [8ff3b77](https://github.com/apache/spark/pull/37843/commits/8ff3b77bcebc1c2dd1bfd26a3391d950790071e2) revert to the previous version, because from the local test, using LinkedHashSet is slower than using ArrayList + HashSet, which may be because LinkedHashSet.toarray is much slower than ArrayList.toarray.

Let's wait for the bench results from GA to verify multiple Java versions

Reply:
OK that's fine, thanks for checking. Whatever seems most efficient

Reply:
The test results of GA are as follows:

- Java 8 : `ArrayList  + HashSet` is 5 ~ 10% faster than `LinkedHashSet `
- Java 11:  `ArrayList  + HashSet` is 20 ~ 40% faster than `LinkedHashSet `
- Java 17:  `ArrayList  + HashSet` is 5 ~ 10% **slower** than `LinkedHashSet `

For this reason, I added a comment, when Java 17 is default, we can consider change to using `LinkedHashSet`, Is that ok?

Or should we change to 

```
if (SystemUtils.isJavaVersionAtLeast("17")) {
  use LinkedHashSet
} else {
  use ArrayList + HashSet
}
```
?

Reply:
I tend to keep the status, which is already more efficient than Arrays.stream api

Reply:
Leave it as is without LinkedHashSet in all cases. Keep it simple

Reply:
OK

Reply:
QQ: Is ordering important here ? Else, simply `uniqueValues.toArray()` ?

Reply:
https://github.com/LuciferYang/spark/runs/8300325922?check_suite_focus=true to check is whether there will be UT failure after change to use `uniqueValues.toArray()`

Reply:
The test results of GA are as follows:

- Java 8 : ArrayList + HashSet is 5 ~ 10% faster than HashSet
- Java 11: ArrayList + HashSet is 5 ~ 10% slower than HashSet
- Java 17: ArrayList + HashSet is 5 ~ 10% slower than HashSet

But there are [2 test failed ](https://github.com/LuciferYang/spark/runs/8300340727?check_suite_focus=true)in `V2PredicateSuite`, the test failure is related to the use of `uniqueValues.toArray()`

https://github.com/apache/spark/blob/443eea97578c41870c343cdb88cf69bfdf27033a/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2PredicateSuite.scala#L266

https://github.com/apache/spark/blob/443eea97578c41870c343cdb88cf69bfdf27033a/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/v2/V2PredicateSuite.scala#L290

Judging from the intention of the assertion, ordering is considered. We may need more people to check this if we want change to use `uniqueValues.toArray()`

Reply:
friendly ping @huaxingao Could you help confirm that is the ordering important for the result of `Expression#references` method?

Reply:
Thanks @huaxingao

Reply:
@mridulm @srowen [46d4a57](https://github.com/apache/spark/pull/37843/commits/46d4a578d8649328372ee785ac74ecadebf17961) change to use HashSet due to the result order is not important

Reply:
ok to use HashSet if the result order is not important